File: MATLAB/run_sshybrid.m
% Author: Sina Hafezi
% Enhancement processing script for SS-Hybrid beamformer
% Paper title: Subspace Hybrid Beamforming for head-worn microphone arrays
% Conference: ICASSP 2023 

% Version history: date/author/updates:
% 2023 Mar - Sina Hafezi - first matlab script to run SS-Hybrid beamforming enhancement

clear
close all;

setup;

% examle data
test_case=1; % test case number
dataPath = fullfile('..','example data',num2str(test_case)); % root path to example data
savePath = 'outputs/';mkdir(savePath); % where to optionally save the output audio
ToSave=0; % whether to save output audio in savePath


% processing setting
fs = 10e3; % sample rate
refChan = 2; % array reference channel
settings.Tau = 80e-3; % 80 ms for PCA smoothing
settings.dict = load("﻿sshybrid_dictionary.mat"); % load dictionary

% load array audio
[y,fs0]=audioread(fullfile(dataPath,'array.wav'));
y=resample(y,fs,fs0);
% load target DOA (wrt. head coordinate system)
tinfo=readcell(fullfile(dataPath,'info.txt'));
target_id=tinfo{find(contains(tinfo(:,1),'Target ID')),2};
target = read_participant_relative_position(fullfile(dataPath,'ht.json'),target_id);
target_pos = zeros(length(target.t),3); %[time(sec) azimuth(deg) elevation(deg]
target_pos(:,1)=target.t(:);
[target_pos(:,2) target_pos(:,3) ~]=cart2sph(target.x(target.t(:)),target.y(target.t(:)),target.z(target.t(:)));
target_pos(:,2:3) = wrapTo180(rad2deg(target_pos(:,2:3)));


% SS-Hybrid beamforming
[z_sshyb,z_iso,z_hyb] = fcn_mvdr_sshybrid(y,fs,target_pos,settings);
z_pass = y(:,refChan); % passthrough signal

% To listen
gain=20;
% sound(gain*z_pass,fs);
% sound(gain*z_iso,fs);
% sound(gain*z_hyb,fs);
% sound(gain*z_sshyb,fs);

% To save outupt audio
if ToSave
    audiowrite(fullfile(savePath,sprintf('test_%d_%s.wav',test_case,'pass')),gain*z_pass,fs);
    audiowrite(fullfile(savePath,sprintf('test_%d_%s.wav',test_case,'iso')),gain*z_iso,fs);
    audiowrite(fullfile(savePath,sprintf('test_%d_%s.wav',test_case,'hyb')),gain*z_hyb,fs);
    audiowrite(fullfile(savePath,sprintf('test_%d_%s.wav',test_case,'sshyb')),gain*z_sshyb,fs);
end


%% STFT plots
CL=[-100 -50];
figure;tiledlayout(4,1,'TileSpacing','tight','Padding','tight');
nexttile;fcn_plot_spec(z_pass,fs);title('Passthrough');clim(CL);
nexttile;fcn_plot_spec(z_iso,fs);title('Iso');clim(CL);
nexttile;fcn_plot_spec(z_hyb,fs);title('Hybrid');clim(CL);
nexttile;fcn_plot_spec(z_sshyb,fs);title('SS-Hybrid');clim(CL);
if ToSave exportgraphics(gcf,fullfile(savePath,sprintf('test_%d.png',test_case))); end

File: Python/run_sshybrid.py
# Author: Sina Hafezi
# Enhancement processing script for SS-Hybrid beamformer
# Paper title: Subspace Hybrid Beamforming for head-worn microphone arrays
# Conference: ICASSP 2023 

# Version history: date/author/updates:
# 2023 Mar - Sina Hafezi - first Python script to run SS-Hybrid beamforming enhancement


from SSHybrid import *
from EasyCom_Array import *
from DataLoader import *
import os
import numpy as np
import sounddevice as sd # only used to play audio
import soundfile as sf # only used to save audio

# settings for processing
fs = int(10e3)
test_case = 1
out_path = 'outputs'
ToSave = True # whether to save the output audio in out_path folder
if not os.path.exists(out_path):
    os.makedirs(out_path) 

# path setup
atf_file = os.path.join('..', 'example data','Device_ATFs.h5')
data_path = os.path.join('..', 'example data',str(test_case))

AIR = EasyCom_Array(atf_file) # get EasyCom array AIR (Acoustic Impulse Response)
sshyb = SSHybrid(AIR, fs) # initialize SS-Hybrid for this array

# (OPTIONAL) get and plot few sample 2D & 3D noise field isotropy models
pow_dyns=np.arange(-8,-48,-8) # dynamic ranges for horizontally unimodal anisotropic noise field models
ps = sshyb.generate_2D_anisotropies(pow_dyns) # 2D isotropies
ps3 = sshyb.convert_2D_to_3D_isotropy(ps) # 3D isotropies (quadrature weighted along elevation)
sshyb.plot_2D_isotropy(ps[:,0,:],sshyb.azi) # plot few examples of 2D isotropies
sshyb.plot_3D_isotropy(ps3[:,0,0], sshyb.azi, sshyb.ele) # plot an example of 3D isotropy

# Loading or Making Dictionary
#sshyb.make_dictionary(pow_dyns=pow_dyns,steer_az_limit=[-30,30],steer_el_limit=[-90,90])
#sshyb.save_dictionary()
sshyb.load_dictionary()

# READ in target DOA (target doas)
dl = DataLoader(data_path)
dl.plot_VAD() # plot voice activity labels
target_doas = dl.get_target_doa() # (azi_deg, ele_deg) [N x 2]
target_doas = np.concatenate((dl.t[:,None],target_doas),axis=1) # (t_sec, azi_deg, ele_deg) [N x 3]

# READ in naudio
x, _ = sshyb.read_audio(os.path.join(data_path,'array.wav'),fs=fs)
y_pass = x[sshyb.out_chan[0],:]

# PROCESS audio
y_sshyb, y_iso, y_hyb = sshyb.process_signal(x,target_doas)

# PLAY output
#sd.play(y_pass.T, fs) # passthrough signal (no processing)
#sd.play(y_iso.T, fs) # superdiredtive baseline enhanced
#sd.play(y_sshyb.T, fs) # enhaned output (SS-Hybrid) 

# Export audio
if ToSave:
    gain=20;
    sf.write(os.path.join(out_path,'case_%d_%s.wav'%(test_case,'pass')), gain*y_pass.T, fs)
    sf.write(os.path.join(out_path,'case_%d_%s.wav'%(test_case,'iso')), gain*y_iso.T, fs)
    sf.write(os.path.join(out_path,'case_%d_%s.wav'%(test_case,'hyb')), gain*y_hyb.T, fs)
    sf.write(os.path.join(out_path,'case_%d_%s.wav'%(test_case,'sshyb')), gain*y_sshyb.T, fs)

# PLOT STFTs
sshyb.plot_stft_sig(y_pass,'Passthrough')
sshyb.plot_stft_sig(y_iso,'Iso MVDR (Superdirective)')
sshyb.plot_stft_sig(y_hyb,'Hybrid MVDR (containing musical noise)')
sshyb.plot_stft_sig(y_sshyb,'SS-Hybrid MVDR (proposed)')

